### PR TITLE
Don't chain imports with *

### PR DESCRIPTION
**Pull request by esoergel**
From the slack convo yesterday.
@gcapalbo

### CODE REVIEW

File: corehq/apps/users/models.py [hunk 1]
```diff
@@ -2548,7 +2550,3 @@ def get(self, user_id):
             user = CouchUser.get_by_user_id(user_id)
             self.cache[user_id] = user
             return user
-
-
-from .signals import *
```
Comment (czue) [inline corehq/apps/users/models.py]: I think the (very long time ago) purpose of this import was to make sure that all the `signal.connect` code fired whenever django bootstrapped. haven't read slack convo yet, but wanted to confirm that we were confident this change didn't accidentally change that behavior

Comment (gcapalbo) [inline corehq/apps/users/models.py]: that's a good point, it looks like signals.py has some code it executes to setup post-save signals which now won't get setup until the import happens within the function calls above

Comment (czue) [inline corehq/apps/users/models.py]: it looks like we should replace this with an `AppConfig`. you can see @snopoke's recent change to fluff as an example: https://github.com/dimagi/fluff/pull/105

Comment (snopoke) [inline corehq/apps/users/models.py]: See [Where should this code live?](https://docs.djangoproject.com/en/1.8/topics/signals/#connecting-receiver-functions)

File: corehq/apps/users/__init__.py [hunk 2]
```diff
@@ -1,4 +1,17 @@
...
-ExtraPreindexPlugin.register('users', __file__, settings.NEW_USERS_GROUPS_DB)
+from .signals import connect_user_signals
+
+
+class UsersAppConfig(AppConfig):
+    name = 'corehq.apps.users'
+
+    def ready(self):
+        """Code to run with Django starts"""
+        ExtraPreindexPlugin.register('users', __file__, settings.NEW_USERS_GROUPS_DB)
```
Comment (esoergel) [inline corehq/apps/users/__init__.py]: @dannyroberts FYI - I took the liberty of moving this into `AppConfig.ready()`.  Know of any reason why that's not a good idea?

File: corehq/apps/users/__init__.py [hunk 3]
```diff
@@ -1,4 +1,17 @@
...
+        connect_user_signals()
+
+
+default_app_config = 'corehq.apps.users.UsersAppConfig'
```
Comment (esoergel) [inline corehq/apps/users/__init__.py]: According to the docs, you can just change the `INSTALLED_APPS` entry to point to this class, but I kept getting a weird, cryptic error when I tried that - I think we do other things with `INSTALLED_APPS`, and changing this broke something non-obvious.  This seems to work, though.

File: corehq/apps/users/signals.py [hunk 4]
```diff
@@ -11,6 +11,10 @@
 from casexml.apps.phone.xml import VALID_USER_REGISTRATION_XMLNSES
 
 
+commcare_user_post_save = Signal(providing_args=["couch_user"])
+couch_user_post_save = Signal(providing_args=["couch_user"])
```
Comment (esoergel) [inline corehq/apps/users/signals.py]: It felt like these two should be right at the top of the file

**Comment by gcapalbo** [comment]
:+1: Will give others a chance to review before merge as well, but looks good

Do you know if there's test coverage on the things that connect to those signals?

**Comment by esoergel** [comment]
@gcapalbo I do not, but I verified that the signal handlers are being
called on save

On Tue, Oct 20, 2015 at 4:35 PM, gcapalbo <notifications@github.com> wrote:

> [image: :+1:] Will give others a chance to review before merge as well,
> but looks good
>
> Do you know if there's test coverage on the things that connect to those
> signals?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/dimagi/commcare-hq/pull/8693#issuecomment-149694221>.
>



--

File: corehq/apps/users/__init__.py (hunk 2 above)
Comment (dannyroberts) [inline corehq/apps/users/__init__.py]: Off the top of my head, this looks fine.

**Comment by czue** [comment]
adding tests for each signal sounds like it'd be easy and helpful

File: corehq/apps/users/models.py [hunk 5]
```diff
@@ -1414,7 +1416,7 @@ def clear_quickcache_for_user(self):
     def save(self, **params):
         super(CommCareUser, self).save(**params)
 
-        from corehq.apps.users.signals import commcare_user_post_save
+        from .signals import commcare_user_post_save
```
Comment (snopoke) [inline corehq/apps/users/models.py]: Why the change to relative import?

Comment (esoergel) [inline corehq/apps/users/models.py]: No particular reason - I was adding the same import before each of the signal connections, so I just made them consistent, and I generally prefer explicit relative imports for stuff within the same module.  Actually, I could probably just import this at the top of the file now that the signal binding has moved to the AppConfig.

**Comment by esoergel** [comment]
Turns out testing signals was riddled with confusing gotchas.  I finally got it working though, and I think I understand the problems, so feel free to ping me if that's something you're trying to do in the future.  I also left some comments in the test file.

As a bonus, I added a [`FakeCouch` decorator](https://github.com/dimagi/commcare-hq/commit/4ec66e80ffdcb83306d516068ff4bad58d74201a)

**Comment by gcapalbo** [comment]
Thanks @esoergel for putting that together.  I think the build failure is real though.

**Comment by czue** [comment]
(kicked build)
